modules/tracking/src/trackerKCF.cpp

```diff
@@ -0,0 +1,538 @@
+	}
+      }
+    }
+    
+    cnFeatures=temp.clone();
+  }
+  
+  /*
+   *  dense gauss kernel function
+   */
+  void TrackerKCFImpl::denseGaussKernel(const double sigma, const Mat _x, const Mat _y, Mat & _k)const{
+    Mat _xf, _yf, xyf,xy;
+    double normX, normY;
+    
+    fft2(_x,_xf);
+    fft2(_y,_yf);
+    normX=norm(_x);
+    normX*=normX;
+    normY=norm(_y);
+    normY*=normY;
+    
+    mulSpectrums(_xf,_yf,xyf,0,true);
+      
+    ifft2(xyf,xyf);
+    shiftRows(xyf, _x.rows/2);
+    shiftCols(xyf,_x.cols/2);
+
+    //(xx + yy - 2 * xy) / numel(x)
+    xy=(normX+normY-2*xyf)/(_x.rows*_x.cols);
+
+
+    // TODO: check wether we really need thresholding or not
+    //threshold(xy,xy,0.0,0.0,THRESH_TOZERO);//max(0, (xx + yy - 2 * xy) / numel(x))
+    for(int i=0;i<xy.rows;i++){
+      for(int j=0;j<xy.cols;j++){
+	if(xy.at<double>(i,j)<0.0)xy.at<double>(i,j)=0.0;
+      }
+    }
+    
+    double sig=-1.0/(sigma*sigma);
+    xy=sig*xy;
+    exp(xy,_k);
+
+  }
+  
+  /* CIRCULAR SHIT Function
+   * http://stackoverflow.com/questions/10420454/shift-like-matlab-function-rows-or-columns-of-a-matrix-in-opencv
+   */
+  // circular shift one row from up to down
+  void TrackerKCFImpl::shiftRows(Mat& mat) const {
+
+      Mat temp;
+      Mat m;
+      int _k = (mat.rows-1);
+      mat.row(_k).copyTo(temp);
+      for(; _k > 0 ; _k-- ) {
+	  m = mat.row(_k);
+	  mat.row(_k-1).copyTo(m);
+      }
+      m = mat.row(0);
+      temp.copyTo(m);
+
+  }
+
+  // circular shift n rows from up to down if n > 0, -n rows from down to up if n < 0
+  void TrackerKCFImpl::shiftRows(Mat& mat, int n) const {
+
+      if( n < 0 ) {
+
+	  n = -n;
+	  flip(mat,mat,0);
+	  for(int _k=0; _k < n;_k++) {
+	      shiftRows(mat);
+	  }
+	  flip(mat,mat,0);
+
+      } else {
+
+	  for(int _k=0; _k < n;_k++) {
+	      shiftRows(mat);
+	  }
+      }
+
+  }
+
+  //circular shift n columns from left to right if n > 0, -n columns from right to left if n < 0
+  void TrackerKCFImpl::shiftCols(Mat& mat, int n) const {
+
+      if(n < 0){
+
+	  n = -n;
+	  flip(mat,mat,1);
+	  transpose(mat,mat);
+	  shiftRows(mat,n);
+	  transpose(mat,mat);
+	  flip(mat,mat,1);
+
+      } else {
+
+	  transpose(mat,mat);
+	  shiftRows(mat,n);
+	  transpose(mat,mat);
+      }
+  }
+
+  /*
+   * calculate the detection response
+   */
+  void TrackerKCFImpl::calcResponse(const Mat _alphaf, const Mat _k, Mat & _response)const {
+    //alpha f--> 2channels ; k --> 1 channel; 
+    Mat _kf;
+    fft2(_k,_kf);
+    Mat spec;
+    mulSpectrums(_alphaf,_kf,spec,0,false);
+    ifft2(spec,_response);
+  }
+  /*----------------------------------------------------------------------*/
+  
+  /*
+ * Parameters
+ */
+  TrackerKCF::Params::Params(){
+      sigma=0.2;
+      lambda=0.01;
+      interp_factor=0.075;
+      output_sigma_factor=1.0/16.0;
+      resize=true;
+      max_patch_size=80*80;
+      descriptor=GRAY;
+  }
+
+  void TrackerKCF::Params::read( const cv::FileNode& /*fn*/ ){
```

Comment:
Minor suggestion. In this function and the one below, if you plan to leave their implementation empty, then I suggest to finish them with {} in one line (easier to grep). Otherwise, write TODO or something equivalent.